File: src/main/java/io/diekema/dingo/feast/sources/PipelineAggregateSource.java
package io.diekema.dingo.feast.sources;

import io.diekema.dingo.feast.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

/**
 * Created by rdiekema on 9/23/16.
 */
public class PipelineAggregateSource implements Source {

    private static final Logger log = LoggerFactory.getLogger(PipelineAggregateSource.class.getName());

    private Pipeline[] sourcePipelines;
    private final CompletionService<List<Asset>> completionService;

    public PipelineAggregateSource(Pipeline... sourcePipelines) {
        this.sourcePipelines = sourcePipelines;
        this.completionService = new ExecutorCompletionService<>(Executors.newFixedThreadPool(sourcePipelines.length));
    }

    @Override
    public Exchange collect() throws IOException {

        List<Future<List<Asset>>> submitted = new ArrayList<>();
        for (Pipeline p : sourcePipelines) {
            submitted.add(completionService.submit(new PipelineCallable(p)));
        }

        List<Asset> assetList = new ArrayList<>();
        for(int i = 0; i < submitted.size(); i++){
            try {
                final Future<List<Asset>> result = completionService.take();
                assetList.addAll(result.get());
            } catch (InterruptedException | ExecutionException e) {
                log.error(e.getMessage());
            }
        }

        Exchange exchange = new Exchange();
        exchange.setAssets(assetList);
        return exchange;
    }
}

File: src/main/kotlin/io/diekema/dingo/feast/sources/PipelineAggregateSource.kt
/*
 * Copyright (c) 2016. Richard Diekema - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited Proprietary and confidential
 * Written by Richard Diekema <rdiekema@gmail.com>, September 2016
 */

package io.diekema.dingo.feast.sources

import io.diekema.dingo.feast.Asset
import io.diekema.dingo.feast.Exchange
import io.diekema.dingo.feast.Pipeline
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.*

/**
 * Created by rdiekema on 9/23/16.
 */
class PipelineAggregateSource(private val sourcePipelines: Array<Pipeline>) : Source {

    @Throws(IOException::class)
    override fun collect(): Exchange {
        val assetList = ArrayList<Asset>()

        sourcePipelines
                .map { pipe ->
                    assetList.addAll(pipe.run())
                }


        val exchange = Exchange()
        exchange.assets = assetList
        return exchange
    }

    companion object {

        private val log = LoggerFactory.getLogger(PipelineAggregateSource::class.java.name)
    }
}
